src/mat/interface/permonmatnullspace.c

```diff
@@ -0,0 +1,69 @@
+
+#include <permon/private/permonmatimpl.h>
+
+PetscErrorCode MatSetNullSpaceMat(Mat mat, Mat R)
+{
+  PetscErrorCode ierr;
+
+  PetscFunctionBegin;
+  PetscValidHeaderSpecific(mat,MAT_CLASSID,1);
+#if defined(PETSC_USE_DEBUG)
+  if (R) {
+    PetscValidHeaderSpecific(R,MAT_CLASSID,2);
+    TRY( MatCheckNullSpaceMat(mat, R, PETSC_DEFAULT) );
+  }
+#endif
+  ierr = PetscObjectCompose((PetscObject)mat, "NullSpace_Mat", (PetscObject)R);CHKERRQ(ierr);
+  PetscFunctionReturn(0);
+}
+
+PetscErrorCode MatGetNullSpaceMat(Mat mat, Mat *R)
+{
+  PetscErrorCode ierr;
+
+  PetscFunctionBegin;
+  PetscValidHeaderSpecific(mat,MAT_CLASSID,1);
+  PetscValidPointer(R,2);
+  ierr = PetscObjectQuery((PetscObject)mat, "NullSpace_Mat", (PetscObject*)R);CHKERRQ(ierr);
+  PetscFunctionReturn(0);
+}
+
+PetscErrorCode MatCopyNullSpaceMat(Mat mat1, Mat mat2)
+{
+  Mat R;
+  PetscErrorCode ierr;
+
+  PetscFunctionBegin;
+  PetscValidHeaderSpecific(mat1,MAT_CLASSID,1);
+  PetscValidHeaderSpecific(mat2,MAT_CLASSID,2);
+  ierr = PetscObjectQuery((PetscObject)mat1, "NullSpace_Mat", (PetscObject*)&R);CHKERRQ(ierr);
+  ierr = PetscObjectCompose((PetscObject)mat2, "NullSpace_Mat", (PetscObject)R);CHKERRQ(ierr);
+  PetscFunctionReturn(0);
+}
+
+PetscErrorCode MatCheckNullSpaceMat(Mat K,Mat R,PetscReal tol)
+{
+  Vec Rx,x,y;
+  PetscReal normy;
+
+  PetscFunctionBegin;
+  if (tol == PETSC_DECIDE || tol == PETSC_DEFAULT) tol = 1e2*PETSC_SQRT_MACHINE_EPSILON;
```

Comment:
```suggestion
  if (tol == PETSC_DECIDE || tol == PETSC_DEFAULT) tol = 1e2*PETSC_MACHINE_EPSILON;
```

Reply:
This has already been discussed in https://github.com/permon/permon/commit/09c7296b7a2f8b8a2e6860182dbdb28ba8824848#r39688564

>>> Let's test the absolute criterium - it should be always correct. I would suggest the default tolerance to be 100*eps.
>>
>> Do you mean eps = PETSC_SQRT_MACHINE_EPSILON? That's what MatNullSpaceTest uses.
>
>yes

See also https://gitlab.com/petsc/petsc/-/blob/master/src/mat/interface/matnull.c#L409

Reply:
OK, my mistake in the previous discussion. I meant eps = PETSC_MACHINE_EPSILON. Using 1.49e-5 in double is extremely weak.

Reply:
Where did you dig this number?

https://gitlab.com/petsc/petsc/-/blob/master/include/petscmath.h#L777

Reply:
*e-6, sorry

Reply:
Got you, you already multiplied by the given factor but still it is 1.49e-6, right?

Reply:
I'm reluctant to use stronger criterion than `MatNullSpaceTest` since we even throw an error.

Reply:
Do not be so fixated on what is done in PETSc. The important thing is why should we use such a weak tolerance. Using 1.49e-6 allows for perturbations in the nullspace of order ~1e-6 (i.e. 3x3 Laplacian with nullspace [1+1e-6;1;1] satisfies the tolerance). I think that 1e-8 (sqrt_eps) should be an absolute minimum for the tolerance (which coincidently, I believe, is the threshold for pivots when computing the nullspace).

Reply:
> Do not be so fixated on what is done in PETSc. 

I am, because my overall plan this functionality should be in PETSc anyway. If I retrospect how much time I spent with this PERMON PR, I think I should have done that right away.

> Using 1.49e-6 allows for perturbations in the nullspace of order ~1e-6 (i.e. 3x3 Laplacian with nullspace [1+1e-6;1;1] satisfies the tolerance). I think that 1e-8 (sqrt_eps) should be an absolute minimum for the tolerance (which coincidently, I believe, is the threshold for pivots when computing the nullspace).

PETSc has 10*sqrt_eps ~ 1e-7, not 100*sqrt_eps [which I originally thought you were proposing]. I'm OK with sqrt_eps.

Reply:
Changed to sqrt_eps in 705fc45. Closing.